rust/kernel/static_call.rs

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: GPL-2.0
+
+// Copyright (C) 2024 Google LLC.
+
+//! Logic for static calls.
+
+#[macro_export]
+#[doc(hidden)]
+macro_rules! to_underscore {
+    ($arg:expr) => {
+        _
+    };
+}
+
+/// Imports the given static call.
+///
+/// To [`static_call!`] a function, you must import it into the local scope using this macro.
+///
+/// [`static_call!`]: macro@static_call
+#[macro_export]
+macro_rules! use_static_call {
+    ($name:ident) => {
+        $crate::macros::paste! { use $crate::bindings:: [<__SCT__ $name >]; }
+        $crate::macros::paste! { use $crate::bindings:: [<__SCK__ $name >]; }
+    };
+}
+
+#[doc(hidden)]
+pub struct AddressableStaticCallKey {
+    _ptr: *const bindings::static_call_key,
+}
+unsafe impl Sync for AddressableStaticCallKey {}
+impl AddressableStaticCallKey {
+    pub const fn new(ptr: *const bindings::static_call_key) -> Self {
+        Self { _ptr: ptr }
+    }
+}
+
+#[cfg(CONFIG_HAVE_STATIC_CALL)]
+#[doc(hidden)]
+#[macro_export]
+macro_rules! _static_call {
+    ($name:ident($($args:expr),* $(,)?)) => {{
+        // Symbol mangling will give this symbol a unique name.
+        #[cfg(CONFIG_HAVE_STATIC_CALL_INLINE)]
+        #[link_section = ".discard.addressable"]
+        #[used]
+        static __ADDRESSABLE: $crate::static_call::AddressableStaticCallKey = unsafe {
+            $crate::static_call::AddressableStaticCallKey::new(::core::ptr::addr_of!(
+                $crate::macros::concat_idents!(__SCK__, $name)
+            ))
+        };
+
+        let fn_ptr: unsafe extern "C" fn($($crate::static_call::to_underscore!($args)),*) -> _ =
```

Comment:
Maybe we should use `macro_metavar_expr` feature to avoid this.

Then we can do
```rust
let fn_ptr: unsafe extern "C" fn($(${ignore($args)}_),*) -> _ =
```

Reply:
It doesn't feel like the benefit is worth the trouble of a dependency on extra unstable feature.

Reply:
I'm keeping it as-is.